src/common/libsubprocess/local.c

```diff
@@ -438,11 +490,30 @@ static int start_local_watchers (flux_subprocess_t *p)
 
     c = zhash_first (p->channels);
     while (c) {
+        if (c->flags & CHANNEL_WRITE)
+            write_credits = true;
         flux_watcher_start (c->buffer_write_w);
         flux_watcher_start (c->buffer_read_w);
         c->buffer_read_w_started = true;
         c = zhash_next (p->channels);
     }
+
+    if (p->ops.on_credit && write_credits) {
+        if (!(p->flags & FLUX_SUBPROCESS_FLAGS_NO_INITIAL_CREDITS)) {
+            /* to send initial credits the size of each channel buffer */
+            p->initial_credits_w = flux_prepare_watcher_create (p->reactor,
+                                                                initial_credits_cb,
+                                                                p);
+            if (!p->initial_credits_w) {
+                llog_debug (p,
+                            "flux_prepare_watcher_create: %s",
+                            strerror (errno));
+                return -1;
+            }
+            flux_watcher_start (p->initial_credits_w);
+        }
```

Comment:
Maybe we should eliminate the FLUX_SUBPROCESS_FLAGS_NO_INITIAL_CREDIT flag and _never_ make the initail `on_credit()` callback.  A subprocess user (local or remote) should already know the size of all their buffers.  Either it's SUBPROCESS_DEFAULT_BUFSIZE (now in a public heador) or they explicitly set it.

There is a fair bit of complexity here just to tell the user what they should already know.

Reply:
Oh well, SUBPROCESS_DEFAULT_BUFSIZE doesn't apply to sdexec but we could fix that.

Reply:
My first implementation did exactly what you just described, but I felt the "initial credits" was ultimately important from a programming perspective.

Many times `flux_subprocess_write()` will be called within the `on_credit()` callback.  So we need something to "kickstart" writing to stdin, and the initial credits is what "kickstarts" it.  Otherwise `on_credit()` will never be called.

Reply:
The main case I'm always thinking of is copying from a file descriptor to a subprocess.  In that case, it seems like you'd probably have an fd  watcher that you'd stop and start based on credit availability and do the writes from there?  This proposal wouldn't affect that case much - just whether the fd watcher begins started or has to wait for a callback to start.

Maybe an example would help?

Reply:
> Maybe an example would help?

Take a look at the unit tests I wrote in `libsubprocess/test/stdio.c`, that sort of illustrates the two styles.  One where all the writing is within the credit callback and the other where the first write is outside of the credit callback b/c the NO_INITIAL_CREDIT flag is set.  The second case requires calling `flux_subprocess_write()` in two places.   Which I guess is why I thought sending the initial credits made things easier, the writing is just in one place.

With your use case example, we'd call `flux_watcher_start()` in two places vs one.